File: symmetric-tree/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        '''
        Determines if the binary tree is a mirror of itself
        Returns true if it is, otherwise false
        '''
        return self.checkMirror(root, root)
    
    def checkMirror(self, root1: TreeNode, root2: TreeNode) -> bool:
        if root1 is None and root2 is None: #'None' is symmetric
            return True
        if root1 is not None and root2 is not None: #Something exists
            if root1.val == root2.val: #Except root, check symmetry
                return self.checkMirror(root1.right, root2.left) and self.checkMirror(root1.left, root2.right)
        return False

File: symmetric-tree/symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    public boolean isSymmetric(TreeNode root) 
    {
        if (root == null) // No tree, is symmetric
            return true;
        
        return DFS(root.left, root.right);
    }
    
    public boolean DFS(TreeNode leftNode, TreeNode rightNode)
    {
        if (leftNode == null && rightNode == null) // Both nodes are null
            return true;
        else if (leftNode == null || rightNode == null) // One node is null
            return false;
        else if (leftNode.val != rightNode.val) // Vals don't match
            return false;
        
        
        // DFS like node structure
        return DFS(leftNode.left, rightNode.right) && DFS(leftNode.right, rightNode.left);
    }
}